### PR TITLE
Add benchmark for opentelemetry-appender-tracing

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #
## Changes

Added basic benchmarks as of now for the below scenarios with results:
  - `log_no_subscriber` - No subscriber registered - `312.83 ps`
  - `log_subscriber_log_level_disabled`  - Subscriber with event_enabled false - `15.016 ns`    
       // tokio::event_enabled() -> processor::event_enabled() -> exporter::event_enabled()
  - `log_subscriber_log_level_enabled`  - Subscriber with event_enabled true -  `380.09 ns`   
        // tokio::event_enabled() -> processor::event_enabled() -> exporter::event_enabled() 
       //+ tracing::info!() -> processor::emit() -> exporter::export()

image
## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
I ❤️ it that we start talking in terms of nanosecs and now picosecs!

File: opentelemetry-appender-tracing/benches/logs.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        name: &str,
+    ) -> bool {
+        self.exporter.event_enabled(level, target, name)
+    }
+}
+
+fn benchmark_no_subscriber(c: &mut Criterion) {
+    c.bench_function("log_no_subscriber", |b| {
+        b.iter(|| {
+            info!("my-event-name");
```
Comment (cijothomas) [inline opentelemetry-appender-tracing/benches/logs.rs:66]: to more closely match practical uses, lets add name/target + 3-5 fields ?

File: opentelemetry-appender-tracing/benches/logs.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                info!("my-event-name");
+            });
+        });
+    });
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    benchmark_no_subscriber(c);
+    benchmark_with_subscriber(c, false, "log_subscriber_disabled");
+    benchmark_with_subscriber(c, true, "log_subscriber_enabled");
```
Comment (cijothomas) [inline opentelemetry-appender-tracing/benches/logs.rs:98]: I think it'd be good to add a tracing-subscriber itself, that is noop, and returns false for is_enabled(). So we can truly compare - what is the overhead introduced by OTel when not_enabled! It should be negligible as we copy level/target and instrumentation name, and pass refs to it to processor..
(https://github.com/open-telemetry/opentelemetry-rust/blob/main/stress/src/logs.rs#L71-L74)

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/logs/log_emitter.rs#L210 This clone needs some rethinking. Currently, every processor, even the ones that can stream event directly will have to suffer from this clone cost. I think we should just pass ref to logrecord to processors, and if a processor wants to batch it, then it can do the clone and pay the cost. This should give a significant cost reduction.

File: opentelemetry-appender-tracing/benches/logs.rs (hunk 1 above)
Comment (lalitb) [reply opentelemetry-appender-tracing/benches/logs.rs]: yes done.

File: opentelemetry-appender-tracing/benches/logs.rs (hunk 2 above)
Comment (lalitb) [reply opentelemetry-appender-tracing/benches/logs.rs]: Good point, added now.

**Comment by lalitb** [comment]
> https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/logs/log_emitter.rs#L210 This clone needs some rethinking. Currently, every processor, even the ones that can stream event directly will have to suffer from this clone cost. I think we should just pass ref to logrecord to processors, and if a processor wants to batch it, then it can do the clone and pay the cost. This should give a significant cost reduction.

#1650

**Comment by lalitb** [comment]
The noop layer's benchmark test shows it takes about 12.5 nanoseconds when is_enabled off. This seems to be the smallest delay we can't avoid for our otel subscriber, even after making it as fast as possible. One option is to register the subscriber only when the opentelemetry-sdk is available, through feature-flag, as @TommyCpp suggested, to make things faster for applications that don't use otel-sdk for logging.

**Review by cijothomas — APPROVED** [review]

**Comment by lalitb** [comment]
Also checked-in the benchmark results for record. It's beneficial to keep these for future comparisons. Please let me know if there are any issues.

**Comment by cijothomas** [comment]
> Also checked-in the benchmark results for record. It's beneficial to keep these for future comparisons. Please let me know if there are any issues.

yes, I personally love keeping some numbers for quick reference!

**Pull request merged**